File: lib/src/main/java/io/github/kory33/guardedqueries/core/utils/extensions/MapExtensions.java
package io.github.kory33.guardedqueries.core.utils.extensions;

import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSet;

import java.util.Collection;
import java.util.Map;
import java.util.function.Function;

public class MapExtensions {
    /**
     * Computes a map that maps each value in {@code values} to its preimage in {@code map}.
     */
    public static <K, V> ImmutableMap<V, /* possibly empty, disjoint */ImmutableSet<K>> preimages(
            final Map<K, V> map,
            final Collection<V> values
    ) {
        final var valueSet = ImmutableSet.copyOf(values);

        return ImmutableMapExtensions.consumeAndCopy(
                valueSet.stream()
                        .map(value -> {
                            final var preimageIterator = map.entrySet().stream()
                                    .filter(entry -> entry.getValue().equals(value))
                                    .map(Map.Entry::getKey)
                                    .iterator();

                            return Map.entry(value, ImmutableSet.copyOf(preimageIterator));
                        })
                        .iterator()
        );
    }

    public static <K, V1, V2> ImmutableMap<K, V2> composeWithFunction(
            final Map<K, V1> map,
            final Function<? super V1, ? extends V2> function
    ) {
        return ImmutableMapExtensions.consumeAndCopy(
                map.entrySet().stream()
                        .map(entry -> Map.entry(entry.getKey(), function.apply(entry.getValue())))
                        .iterator()
        );
    }
}

File: lib/src/main/scala/io/github/kory33/guardedqueries/core/utils/extensions/MapExtensions.scala
package io.github.kory33.guardedqueries.core.utils.extensions

import io.github.kory33.guardedqueries.core.utils.datastructures.BijectiveMap

import java.util
import java.util.stream.Stream
import scala.jdk.CollectionConverters.*

object MapExtensions {

  /**
   * Computes a map that maps each value in {@code values} to its preimage in {@code map}.
   */
  def preimages[K, V](map: Map[K, V], values: Set[V]): Map[V, Set[K]] =
    map.groupMap(_._2)(_._1)
      .view
      .mapValues(_.toSet)
      .filterKeys(values.contains)
      .toMap

  def restrictToKeys[K, V](map: Map[K, V], keys: Set[K]): Map[K, V] =
    map.view.filterKeys(keys.contains).toMap

  def restrictToKeys[K, V](map: BijectiveMap[K, V], keys: Set[K]): BijectiveMap[K, V] =
  // this call to .get never throws since a restriction of an injective map is again injective
    BijectiveMap.tryFromInjectiveMap(restrictToKeys(map, keys)).get
}
